pkg/engine/plugins.go

```diff
@@ -43,6 +43,80 @@ const (
 	preparePluginVerboseLog = 8
 )
 
+// PackageSet represents a set of packages.
+type PackageSet map[string]workspace.PackageDescriptor
+
+// Add adds a package to this package set.
+func (p PackageSet) Add(plug workspace.PackageDescriptor) {
+	p[plug.String()] = plug
+}
+
+// NewPackageSet creates a new empty pluginSet.
+func NewPackageSet(plugins ...workspace.PackageDescriptor) PackageSet {
+	var s PackageSet = make(map[string]workspace.PackageDescriptor, len(plugins))
+	for _, p := range plugins {
+		s.Add(p)
+	}
+	return s
+}
+
+// Union returns the union of this PackageSet with another PackageSet.
+func (p PackageSet) Union(other PackageSet) PackageSet {
+	newSet := NewPackageSet()
+	for _, value := range p {
+		newSet.Add(value)
+	}
+	for _, value := range other {
+		newSet.Add(value)
+	}
+	return newSet
+}
+
+// ToPluginSet converts this PackageSet to a PluginSet by discarding all parameterization information.
+func (p PackageSet) ToPluginSet() PluginSet {
+	newSet := NewPluginSet()
+	for _, value := range p {
+		newSet.Add(value.PluginSpec)
+	}
+	return newSet
+}
+
+// A PackageUpdate represents an update from one version of a package to another.
+type PackageUpdate struct {
+	// The old package version.
+	Old workspace.PackageDescriptor
+	// The new package version.
+	New workspace.PackageDescriptor
+}
+
+// UpdatesTo returns a list of PackageUpdates that represent the updates to the argument PackageSet present in this
+// PackageSet. For instance, if the argument contains a package P at version 3, and this PackageSet contains the same
+// package P (as identified by name and kind) at version 5, this method will return an update where the Old field
+// contains the version 3 instance from the argument and the New field contains the version 5 instance from this
+// PackageSet.
```

Comment:
Should probably tweak this slightly to mention that we'll consider parameterization when checking equality too.